File: GAMECARO/Network/ConnectToServer.cs
using Client.Game;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Windows.Forms;

namespace Client.Network
{
    public class ConnectToServer
    {
        private readonly string _ip;
        private readonly int _port;
        private Socket? _client;
        private CancellationTokenSource? _cts;
        private readonly StringBuilder _buf = new();
        private readonly Board _board;

        public string PlayerId { get; private set; } = "";
        public string Mark { get; private set; } = "?";

        public event Action? OnBoardChanged;
        public event Action<string>? OnWinner;
        public event Action<string>? OnInitReceived;
        public event Action<int, string>? OnTimerUpdate;
        public event Action? OnReset;
        public event Action<string, string>? OnChatReceived;


        // thêm event rank
        public event Action<int, int, int>? OnRankUpdate;

        public ConnectToServer(string ip, int port, Board board)
        {
            _ip = ip;
            _port = port;
            _board = board;
        }

        public void ConnectAndJoin(string name)
        {
            var ep = new IPEndPoint(IPAddress.Parse(_ip), _port);
            _client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            _client.Connect(ep);
            _cts = new CancellationTokenSource();
            _ = Task.Run(() => Listen(_cts.Token));

            SendJson(new { Action = "JOIN", Name = name });
        }

        private async Task Listen(CancellationToken ct)
        {
            var buffer = new byte[2048];
            while (!ct.IsCancellationRequested && _client != null && _client.Connected)
            {
                int n = await _client.ReceiveAsync(buffer, SocketFlags.None);
                if (n == 0) break;

                var chunk = Encoding.UTF8.GetString(buffer, 0, n);
                _buf.Append(chunk);

                while (true)
                {
                    var cur = _buf.ToString();
                    int idx = cur.IndexOf('\n');
                    if (idx < 0) break;

                    var line = cur[..idx].Trim();
                    _buf.Remove(0, idx + 1);
                    if (string.IsNullOrWhiteSpace(line)) continue;

                    Handle(line);
                }
            }
        }

        private void Handle(string json)
        {
            using var doc = JsonDocument.Parse(json);
            string action = doc.RootElement.GetProperty("Action").GetString() ?? "";

            switch (action)
            {
                case "INIT":
                    PlayerId = doc.RootElement.GetProperty("PlayerId").GetString() ?? "";
                    Mark = doc.RootElement.GetProperty("Mark").GetString() ?? "?";
                    OnInitReceived?.Invoke(Mark);
                    break;

                case "BOARD":
                    var board = doc.RootElement.GetProperty("Board").Deserialize<int[][]>();
                    var turn = doc.RootElement.GetProperty("NextTurn").GetString() ?? "X";
                    string? winner =
                        doc.RootElement.TryGetProperty("Winner", out var w)
                        && w.ValueKind != JsonValueKind.Null ? w.GetString() : null;

                    if (board != null)
                    {
                        _board.Update(board, turn, winner);
                        OnBoardChanged?.Invoke();
                        if (winner != null)
                            OnWinner?.Invoke(winner);
                    }
                    break;

                case "TIMER":
                    int sec = doc.RootElement.GetProperty("Seconds").GetInt32();
                    string turnNow = doc.RootElement.GetProperty("Turn").GetString() ?? "?";
                    OnTimerUpdate?.Invoke(sec, turnNow);
                    break;

                case "RESET":
                    _board.Reset();
                    OnReset?.Invoke();
                    break;

                // thêm case rank
                case "RANK":
                    int score = doc.RootElement.GetProperty("Score").GetInt32();
                    int wins = doc.RootElement.GetProperty("Wins").GetInt32();
                    int losses = doc.RootElement.GetProperty("Losses").GetInt32();

                    OnRankUpdate?.Invoke(score, wins, losses);
                    break;

                case "NOT_READY":
                    string msg = doc.RootElement.GetProperty("Message").GetString() ?? "Chưa đủ người!";
                    MessageBox.Show(msg, "⚠️ Thông báo", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    break;

                case "CHAT":
                    {
                        string player = doc.RootElement.GetProperty("From").GetString() ?? "Unknown";
                        string chatmsg = doc.RootElement.GetProperty("Message").GetString() ?? "";

                        OnChatReceived?.Invoke(player, chatmsg);
                    }
                    break;

                case "SURRENDER":
                    {
                        var winBySurrender = doc.RootElement.GetProperty("Winner").GetString() ?? "";
                        OnWinner?.Invoke(winBySurrender);
                    }
                    break;
            }
        }

        public void SendMove(int row, int col)
        {
            SendJson(new { Action = "MOVE", PlayerId, X = col, Y = row });
        }

        public void SendChat(string message)
        {
            SendJson(new
            {
                Action = "CHAT",
                PlayerId = PlayerId,
                Message = message
            });
        }// done
        // dau hang
        public void SendSurrender()
        {
            SendJson(new
            {
                Action = "SURRENDER",
                PlayerId = PlayerId
            });
        }// done
        private void SendJson(object o)
        {
            if (_client == null) return;
            string s = JsonSerializer.Serialize(o) + "\n";
            byte[] data = Encoding.UTF8.GetBytes(s);
            try { _client.Send(data); } catch { }
        }

        public void Disconnect()
        {
            try { _cts?.Cancel(); _client?.Shutdown(SocketShutdown.Both); } catch { }
            try { _client?.Close(); } catch { }
        }


    }
}

File: GAMECARO/src/ConnectToServer.cpp
#include "Network/ConnectToServer.h"
#include "Game/Board.h"
using namespace Client::Network;
using namespace Client::Game;

ConnectToServer::ConnectToServer(String^ h, int p, Board^ b)
{
    host = h;
    port = p;
    board = b;
    Mark = "";
    client = nullptr;
    stream = nullptr;
}

void ConnectToServer::ConnectAndJoin(String^ playerName)
{
    try
    {
        client = gcnew TcpClient(host, port);
        stream = client->GetStream();

        // Send join message (simple protocol, e.g., "JOIN:playerName")
        array<Byte>^ data = Encoding::UTF8->GetBytes("JOIN:" + playerName);
        stream->Write(data, 0, data->Length);

        // Start listening thread
        listenThread = gcnew Thread(gcnew ThreadStart(this, &ConnectToServer::ListenServer));
        listenThread->IsBackground = true;
        listenThread->Start();
    }
    catch(Exception^ e)
    {
        Console::WriteLine("Connect error: {0}", e->Message);
    }
}

void ConnectToServer::ListenServer()
{
    array<Byte>^ buffer = gcnew array<Byte>(1024);
    int bytesRead;

    try
    {
        while(client->Connected)
        {
            bytesRead = stream->Read(buffer, 0, buffer->Length);
            if(bytesRead > 0)
            {
                String^ msg = Encoding::UTF8->GetString(buffer, 0, bytesRead);
                ProcessMessage(msg);
            }
        }
    }
    catch(Exception^ e)
    {
        Console::WriteLine("Server disconnected: {0}", e->Message);
        Disconnect();
    }
}

void ConnectToServer::ProcessMessage(String^ msg)
{
    // Very simple protocol parsing (for demo)
    // Example: "BOARD:X:row:col", "WINNER:X", "TIMER:sec:turn", "RESET", "RANK:score:wins:losses", "CHAT:player:message"
    array<String^>^ parts = msg->Split(':');
    if(parts->Length == 0) return;

    String^ cmd = parts[0];

    if(cmd == "BOARD" && parts->Length == 4)
    {
        String^ mark = parts[1];
        int row = Int32::Parse(parts[2]);
        int col = Int32::Parse(parts[3]);

        board->PlaceMove(row, col, mark);

        if(OnBoardChanged != nullptr) OnBoardChanged();
    }
    else if(cmd == "WINNER" && parts->Length == 2)
    {
        if(OnWinner != nullptr) OnWinner(parts[1]);
    }
    else if(cmd == "INIT" && parts->Length == 2)
    {
        Mark = parts[1];
        if(OnInitReceived != nullptr) OnInitReceived(Mark);
    }
    else if(cmd == "TIMER" && parts->Length == 3)
    {
        int sec = Int32::Parse(parts[1]);
        int turn = Int32::Parse(parts[2]);
        if(OnTimerUpdate != nullptr) OnTimerUpdate(sec, turn);
    }
    else if(cmd == "RESET")
    {
        if(OnReset != nullptr) OnReset();
    }
    else if(cmd == "RANK" && parts->Length == 4)
    {
        int score = Int32::Parse(parts[1]);
        int wins = Int32::Parse(parts[2]);
        int losses = Int32::Parse(parts[3]);
        if(OnRankUpdate != nullptr) OnRankUpdate(score, wins, losses);
    }
    else if(cmd == "CHAT" && parts->Length >= 3)
    {
        String^ player = parts[1];
        String^ message = String::Join(":", parts, 2, parts->Length - 2);
        if(OnChatReceived != nullptr) OnChatReceived(player, message);
    }
}

void ConnectToServer::SendMove(int row, int col)
{
    if(client == nullptr || !client->Connected) return;

    String^ msg = String::Format("MOVE:{0}:{1}", row, col);
    array<Byte>^ data = Encoding::UTF8->GetBytes(msg);
    stream->Write(data, 0, data->Length);
}

void ConnectToServer::SendChat(String^ message)
{
    if(client == nullptr || !client->Connected) return;

    String^ msg = "CHAT:" + message;
    array<Byte>^ data = Encoding::UTF8->GetBytes(msg);
    stream->Write(data, 0, data->Length);
}

void ConnectToServer::SendSurrender()
{
    if(client == nullptr || !client->Connected) return;

    String^ msg = "SURRENDER";
    array<Byte>^ data = Encoding::UTF8->GetBytes(msg);
    stream->Write(data, 0, data->Length);
}

void ConnectToServer::Disconnect()
{
    try
    {
        if(stream != nullptr) stream->Close();
        if(client != nullptr) client->Close();
        if(listenThread != nullptr && listenThread->IsAlive) listenThread->Abort();
    }
    catch(...) {}
}
